src/python/pants/backend/python/target_types_rules.py

```diff
@@ -38,36 +38,84 @@
+
+    # Handle deprecated cases #6 and #7, which are the only cases where the `sources` field matters
+    # for calculating the entry point.
+    if not ep_val or ep_val.startswith(":"):
+        binary_source_paths = await Get(
+            Paths, PathGlobs, request.sources.path_globs(FilesNotFoundBehavior.error)
+        )
+        if len(binary_source_paths.files) != 1:
+            instructions_url = "https://www.pantsbuild.org/docs/python-package-goal#creating-a-pex-file-from-a-pex_binary-target"
+            if not ep_val:
+                raise InvalidFieldException(
+                    f"The `{ep_alias}` field is not set for the target {address}. Run "
+                    f"`./pants help pex_binary` for more information on how to set the field or "
+                    f"see {instructions_url}."
+                )
             raise InvalidFieldException(
-                f"The `entry_point` field for the target {request.sources.address} is set to "
-                f"the short-hand value {repr(entry_point_value)}, but the `sources` field is not "
-                "set. Pants requires the `sources` field to expand the entry point to the "
-                f"normalized form `path.to.module:{entry_point_value}`. Please either set the "
-                "`sources` field to exactly one file or use a full value for `entry_point`. See "
+                f"The `{ep_alias}` field for the target {address} is set to the short-hand value "
+                f"{repr(ep_val)}, but the `sources` field is not set. Pants requires the "
+                "`sources` field to expand the entry point to the normalized form "
+                f"`path.to.module:{ep_val}`. Please either set the `sources` field to exactly one "
+                f"file or set `{ep_alias}='my_file.py:{ep_val}'`. See "
                 f"{instructions_url}."
             )
-    entry_point_path = binary_source_paths.files[0]
+        entry_point_path = binary_source_paths.files[0]
+        source_root = await Get(
+            SourceRoot,
+            SourceRootRequest,
+            SourceRootRequest.for_file(entry_point_path),
+        )
+        stripped_source_path = os.path.relpath(entry_point_path, source_root.path)
+        module_base, _ = os.path.splitext(stripped_source_path)
+        normalized_path = module_base.replace(os.path.sep, ".")
+        return ResolvedPexEntryPoint(f"{normalized_path}{ep_val}" if ep_val else normalized_path)
+
+    # Case #1.
+    if ep_val in ("<none>", "<None>"):
+        return ResolvedPexEntryPoint(None)
+
+    path, _, func = ep_val.partition(":")
+
+    # If it's already a module (cases #2 and #3), simply use that. Otherwise, convert the file name
+    # into a module path (cases #4 and #5).
+    if not path.endswith(".py"):
+        return ResolvedPexEntryPoint(ep_val)
+
+    # We don't actually need to use the engine here, but we do this to validate that the file
+    # exists and that it resolves to only one file.
```

Comment:
Nit: This comment sortof disagrees with itself: "It doesn't need to use the engine but it does need to."